src/passes/Metrics.cpp

```diff
@@ -92,6 +93,34 @@ struct Metrics : public WalkerPass<PostWalker<Metrics, UnifiedExpressionVisitor<
         counts["[binary-bytes]"] = writer.tableOfContents.functionBodies[i].size;
         printCounts(std::string("func: ") + func->name.str);
       }
+      // print for each export how much code size is due to it, i.e.,
+      // how much the module could shrink without it.
```

Comment:
To be clear, this is `size(self) + sum(size(x) for x in dominated_by(self))` right?

That is, this is *not* the sum of sizes of everything transitively reachable in the call-graph from this function -- it *doesn't* include transitively reachable children that are *also* reachable from other exported functions?

Reply:
What does `dominated_by` mean?

> That is, this is not the sum of sizes of everything transitively reachable in the call-graph from this function -- it doesn't include transitively reachable children that are also reachable from other exported functions?

Correct. This computes how much the binary could shrink if we removed the export. And we can't remove things still needed by other things, the optimizer won't break the module.

But it's also not just the graph of the things only needed by that export, as removing some things might open up optimization opportunities elsewhere. (For example, removing a function might reduce the number of calls of another function from 2 to 1, which might make it beneficial to inline that single-use function, which it wasn't before.)

Reply:
> What does dominated_by mean?

If A dominates B in a graph with some root R, then all paths from R to B *must* pass through A. In this situation, R = any exported function.

https://en.wikipedia.org/wiki/Dominator_(graph_theory)

It is also used in heap profiling: https://developer.mozilla.org/en-US/docs/Tools/Memory/Dominators_view

Reply:
I see, thanks. So dominance with the roots `R` being all the exported functions is equal to "is kept alive by".

Btw, there are 2 other sources of roots, the start function and the table (assuming the table is imported or exported). Would it be useful to show information on those?

Reply:
The start function: definitely. Unsure about table.

Reply:
Added support for the start function.